### ISSUE TITLE
About CIDEr Optimization

### ISSUE THREAD

**Issue by LONGRYUU** [cvpr-anon/captioning#4]
Hello! Thanks for your work. Your code is quite clear and easy to understand. Thus, I'm doing some experiments based on it.

However, I got some problems while training with CIDEr optimization. When I use self-critical strategy to train my pre-trained model, the score on CIDEr tends to drop about 5 points after the first epoch. And it then costs quite a few epoches for the model to achieve the same score as it does with XE loss. Only after these epoches, the model begins to outperform the pre-trained one.

I checked my code and found that I didn't save the state dict of optimizer while training with XE loss. So when I start to train with self-critical strategy, I just initialize a new optimizer with a learning rate about 2e-5 or 5e-5. Is this the reason why I got the problems described above?

**Comment by fawazsammani** [comment]
Hi @LONGRYUU
Sorry for the late reply. I've been quite busy with work.
It should not matter whether you initialize a new optimizer for self-critical training, or use the state dictionary optimizer for XE to continue in self-critical. If the cider drops at the first epoch, that means there is something wrong. Check your cider cache file and make sure that you are using your word map to create it, and not the one I provided (if you've changed your word map dictionary). Another thing that is worth trying is setting your learning rate for self-critical training to a small value without decay. Example, you can set it to 5e-6 and don't decay it. The gradients of self-critical training are large, and thus if your learning rate is large, it won't perform well.

**Comment by LONGRYUU** [comment]
Thanks for your reply! 

Actually, I've tried diffrent learning rates beetween 5e-6 and 5e-5 and found that smaller learning rate sometimes leads to worse performance, so I think maybe there're problems with my cider cache file.
 
I checked my code and noticed that I didn't replace words out of word map with '<unk>'. Besides, I didn't add the '<end>' token in the end of each groudtruth caption.  I'll fix these bugs and see whether my problem can be solved.
